cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/actor/ProgressiveSolver.scala

```diff
@@ -0,0 +1,180 @@
+package edu.uci.ics.cloudberry.zion.actor
+
+import akka.actor.{Actor, ActorLogging, ActorRef, PoisonPill, Props, Stash}
+import akka.util.Timeout
+import edu.uci.ics.cloudberry.zion.TInterval
+import edu.uci.ics.cloudberry.zion.actor.ProgressiveSolver._
+import edu.uci.ics.cloudberry.zion.common.Config
+import edu.uci.ics.cloudberry.zion.model.datastore.IPostTransform
+import edu.uci.ics.cloudberry.zion.model.impl.QueryPlanner.IMerger
+import edu.uci.ics.cloudberry.zion.model.impl.{DataSetInfo, QueryPlanner}
+import edu.uci.ics.cloudberry.zion.model.schema._
+import edu.uci.ics.cloudberry.zion.model.slicing.Drum
+import org.joda.time.DateTime
+import play.api.libs.json.JsArray
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+/**
+  * The Streaming Actor slices one query to a sequence of mini-queries and responds with a stream of partial results.
+  * If it receives multiple queries, the later queries will be stashed until the existing slicing is over.
+  * If it receives a Cancel message, the current slicing will be stopped.
+  *
+  * @param dataManager
+  * @param planner
+  * @param config
+  * @param out
+  */
+class ProgressiveSolver(val dataManager: ActorRef,
+                        val planner: QueryPlanner,
+                        val config: Config,
+                        val out: ActorRef
+                       )(implicit val ec: ExecutionContext) extends Actor with Stash with IQuerySolver with ActorLogging {
+
+  implicit val askTimeOut: Timeout = config.UserTimeOut
+
+  private var ts: Long = 0
+
+  override def receive: Receive = {
+    case request: SlicingRequest =>
+      ts = DateTime.now().getMillis
+      val reporter: ActorRef = context.actorOf(Props(new Reporter(FiniteDuration(request.intervalMS, "ms"), out)))
```

Comment:
Each time a new slicing request session begins, we create a new `reporter` and we kill it when this query finished or it is interrupted by a new incoming request. I am just curious what is the difference if we keep one `reporter` corresponding to the `ProgressiveResolver` and reset it instead of killing and recreating it, like the relation of `ProgressiveResolver` to `BerryClient`.

Reply:
Good point. Changed.